flink-connector-examples/src/main/java/io/pravega/example/flink/streamcuts/process/StreamBookmarker.java

```diff
@@ -0,0 +1,203 @@
+
+    private transient ValueState<StreamCut> startStreamCut;
+    private transient ValueState<StreamCut> lastStreamCutBeforeSlice;
+    private transient ValueState<Tuple2<Integer, Double>> lastValue;
+
+    public Bookmarker(URI pravegaControllerURI) {
+        this.pravegaControllerURI = pravegaControllerURI;
+    }
+
+    @Override
+    public void open(Configuration parameters) {
+        // We will provide Bookmarker with a pointer to see the state of the reader group.
+        startStreamCut = getRuntimeContext().getState(new ValueStateDescriptor<>("startStreamCut",
+                TypeInformation.of(new TypeHint<StreamCut>() {})));
+        lastStreamCutBeforeSlice = getRuntimeContext().getState(new ValueStateDescriptor<>("lastStreamCutBeforeSlice",
+                TypeInformation.of(new TypeHint<StreamCut>() {})));
+        lastValue = getRuntimeContext().getState(new ValueStateDescriptor<>("lastValue",
+                TypeInformation.of(new TypeHint<Tuple2<Integer, Double>>() {})));
+    }
+
+    /**
+     * This function inspects the events written by a particular sensor (keyed stream by sensor id). When this function
+     * finds that there are events < 0 it considers them as of interest for further processing. For this reason, it
+     * gets a StreamCut related to the first observed event < 0 (start of stream slice). Then, if events become > 0
+     * again, the function schedules a timer to get a future StreamCut (end of slice). That is, the SensorStreamSlice
+     * object written to the collector will contain all the events < 0 for a certain sensor.
+     *
+     * @param value Input tuple from a sensor
+     * @param ctx Context
+     * @param out Collector
+     * @throws IOException
+     */
+    @Override
+    public void processElement(Tuple2<Integer, Double> value, Context ctx, Collector<SensorStreamSlice> out) throws IOException {
+        if (lastValue.value() != null) return;
+        // Infer the slice of the stream for which events are < 0.
+        if (value.f1 < 0 && startStreamCut.value() == null) {
+            startStreamCut.update(getReaderGroup().getStreamCuts().get(Stream.of(Constants.DEFAULT_SCOPE, Constants.PRODUCER_STREAM)));
+            LOG.warn("Start bookmarking a stream slice at: {} for sensor {}.", startStreamCut.value(), value.f0);
+        } else if (value.f1 >= 0 && startStreamCut.value() != null && lastStreamCutBeforeSlice.value() == null) {
+            lastStreamCutBeforeSlice.update(getReaderGroup().getStreamCuts().get(Stream.of(Constants.DEFAULT_SCOPE, Constants.PRODUCER_STREAM)));
+            lastValue.update(value);
+            long sliceCreationTime = ctx.timerService().currentProcessingTime() + StreamBookmarker.CHECKPOINT_INTERVAL;
+            LOG.warn("Found events > 0 for sensor {}, waiting for the next stream cut update to create the slice at {}.", value.f0, sliceCreationTime);
+            ctx.timerService().registerProcessingTimeTimer(sliceCreationTime);
```

Comment:
since we reset the stream cut start/end pointers after the timer executes, will we not miss the events (`<0`) that occurred during this interval?

Reply:
The example is prepared in a way that the time interval between the last event `< 0` of a sinus wave and the first event `< 0` of the next one is >> `CHECKPOINT_INTERVAL` (20s vs 5s). This has been a simple assumption to make the code easier to understand by new users. Of course, we could consider more complex scenarios, but it would complicate the logic as well; we can discuss it up to which point this is worth it for this sample.